src/peer_select.cc

```diff
@@ -99,6 +101,124 @@ operator <<(std::ostream &os, const PeerSelectionDumper &fsd)
     return os;
 }
 
+/// An ICP ping timeout service.
+/// Protects event.cc (which is designed to handle a few unrelated timeouts)
+/// from exposure to thousands of ping-related timeouts on busy proxies.
+class PeerSelectorPingMonitor
+{
+public:
+    /// registers the given selector to be notified about the IPC ping timeout
+    void monitor(PeerSelector *);
+
+    /// removes a PeerSelector from the waiting list
+    void forget(PeerSelector *);
+
+    /// \returns a (nil) registration of a non-waiting peer selector
+    WaitingPeerSelectorPosition npos() { return selectors.end(); }
+
+private:
+    static void NoteWaitOver(void *monitor);
+
+    void startWaiting();
+    void abortWaiting();
+    void noteWaitOver();
+
+    WaitingPeerSelectors selectors; ///< \see WaitingPeerSelectors
+};
+
+/// monitors all PeerSelector ICP ping timeouts
+PeerSelectorPingMonitor ThePeerSelectorPingMonitor;
```

Comment:
What happens if, during Squid process termination, ThePeerSelectorPingMonitor object is destroyed _before_ the last PeerSelector is gone? AFAICT, our forget() method may crash during that PeerSelecto object destruction. To be on the safe side, I suggest creating ThePeerSelectorPingMonitor on heap (and never deleting it).

Asserting that ThePeerSelectorPingMonitor is not nil before dereferencing it in startPingWaiting() and cancelPingTimeoutMonitoring() would be safe/OK, but perhaps we should create a PeerSelector::pingMonitor() access method that asserts that ThePeerSelectorPingMonitor is not nil and returns a reference instead.

Reply:
> I suggest creating ThePeerSelectorPingMonitor on heap (and never deleting it).

Done. A new static PingMonitor() now dynamically allocates and returns the reference to the object. No asserts are needed, I think.